### ISSUE TITLE
Idea Signals and Slots communication between objects

### ISSUE THREAD

**Issue by matthiaz** [qcubed/framework#631]
I'd like to share an idea with you guys that might be of use.

As someone who has programmed in QT I have always loved the way they [communicate between objects](http://qt-project.org/doc/qt-4.8/signalsandslots.html). At the company I work, we have tried to emulate this behaviour by altering QControl

```
		abstract class QControl extends QControlBase {
		private $arrCallbacks;
				
		/**
		 * 
		 * @param string $signal callback name (OnSave, OnChange, ...)
		 * @param string $slot name of callback function
		 * @param QControl $receiver qcontrol that has the callback function
		 */
		public function Register($signal, $slot, QControl &$receiver = null) {
			if (!isset($this->arrCallbacks[$signal])) {
				$this->arrCallbacks[$signal] = array();
			}
			if ($receiver) {
				$this->arrCallbacks[$signal][] = array($receiver, $slot);
			} else {
				$this->arrCallbacks[$signal][] = array($this->objParentControl, $slot);
			}
		}

		/**
		 * 
		 * @param string $signal
		 * @param array $args
		 */
		public function Trigger($signal, $args = array()) {
			if (isset($this->arrCallbacks[$signal])) {
				foreach ($this->arrCallbacks[$signal] as $pair) {
					list($receiver, $slot) = $pair;
					if (method_exists($receiver, $slot)) {
						if (count($args) > 0) {
							call_user_func_array(array($receiver, $slot), $args);
						} else {
							$receiver->$slot();
						}
					}
				}
			}
		}
```

The idea is simple. You can add Triggers as much as you want in one class. onSave, onDelete, onFilter, and so on... 
You can then use the Register method to send the callbacks to the correct method. 

The example below will call the function **SetRelationship** in object **$this** when the TableEditor does a **Trigger('OnSave')**
```
$this->TableEditor->Register('OnSave', 'SetRelationship', $this);
```

Is this an idea worth pursuing? Maybe even an implementation in QCubed core? Because it seems odd to make a plugin do this. 
How do other qcubed programmers let their objects communicate with each other? I'm curious, please share.

**Comment by spekary** [comment]
I think this is an idea worth exploring. Its definitely an idea for the core, and not a plugin. However, it could also simply be an idea put online that people implement in their own QControl override class. General concerns on core functionality are supportability, and how much they add to the form state. (This is why I removed the old persistence implementation. Very difficult to support, and bloated the form state).

What you are describing seems quite supportable, and adds little to the form state. The QControl object will get serialized fine. This is essentially a Broadcast implementation, which I remember all the way back in ThinkC was quite useful. 

A couple of thoughts on your code:
* I don't think you need the ampersand in front of the QControl variable. PHP will pass a reference without it. The ampersand would only be needed if you are passing back to the caller a new or different control.
* The title of the functions aren't descriptive enough and clash with the language already in QCubed of triggering actions. I would use something like "Listen", and "Broadcast" (or "Signal"). 
* It occurs to me that you could really easily do this in the QForm, and reduce the amount of code, and not add to the form state at all. The QForm already has an array of all the controls. Why not just have something like the following:

In QForm, have a "Broadcast" function that sends a "signal"(just a constant), to all the controls, by calling the "Signal" function on each control, passing the constant, and maybe even another optional parameter. Controls then implement the signal if they want to, and check the passed constant to see if its the one or ones they care about. You just put a stub into QControlBase that does nothing. Any object in the program can tell the current form (available in the $_FORM global), to broadcast a signal. 

You don't get saved arguments with this method, but I would argue that saving arguments for later use is a bad idea, as the data can go stale and cause problems that are difficult to debug. Remember that this data is going in to the form state, which gets serialized into a database or into a hidden control. If a control really wants to save the data, it can save the arguments in its own private structure. This fixes a potential pitfall in this, in that you have to be really careful about saving a reference to the QForm into the form state. If that happens, the entire form gets serialized twice, and when unserialized, you get two copies of the form, causing major weird problems. The purpose of the Sleep and WakeUp functions sprinkled throughout the core is to prevent that from happening. With the implantation I describe above, the signaler is sending out fresh data.

**Comment by vakopian** [comment]
I like the idea too, and I do remember these from my Qt days. For those who haven't done Qt however, this is similar to the [jQuery on() function](http://api.jquery.com/on/). In fact this comparison tells me that instead of restricting the receiver to a `QControl`, we could hook the signal to a generic callback object (so, in particular the pair `($receiver, $slot)` could be one type of callback). 
Or closer to home, this is not that much different than the Event/Action mechanism already in QCubed, except the events are triggered programmatically rather than automatically from the browser. Could we possibly extend the event/action mechanism to handle the signal/slots use case as well? Like `$control->Trigger(new QSaveEvent())`?

Another idea we might wanna consider to borrow from jQuery (both for signal/slots and event/actions) is the functionality that callbacks registered with `on()` could also trigger for future objects not just the current object the signal is registered on. This can provide big performance benefits: for event/actions we would generate much less javascript code, and for signal/slots we would save on form state.

One more reason why this would be nice to have in core QCubed is that we should have many standard signals generated as part of codegen. For example the `Trigger('OnSave')` call described in @matthiaz example could be codegen-ed to be called from the Panel drafts code, etc.

The only worry I have with introducing these kinds of mechanisms is that it can easily lead to "callback hell". But I guess standard "callback hell" is better than every developer inventing her own callback hell :-).

As a side note about callbacks, I think we should try to standardize different kinds of callbacks used in QCubed under a common interface. I think PHP should have done this a long time ago, but maybe we can do this in user space as well. For example here is my attempt for QCallback class hierarchy: https://github.com/qcubed/framework/blob/feature-jQcubed/includes/app_includes/classes/QCallback.class.php

**Comment by matthiaz** [comment]
@spekary thank you for your feedback. It is not a 100% clear to me how the implementation in QForm that you propose will work. I don't think I'd want to broadcast a signal to all controls if only one or two will actually do something with it. That said: I might have misunderstood. So please do clarify. 
I fully agree that whatever solution is decided on, it shouldn't have an impact on the formstate. Unfortunately, my knowledge of the inner workings of the formstate isn't very advanced. 

@vakopian I like your idea about using actual event objects. But we do have to be careful with standard event objects. QSaveEvent is a pretty standard event but not all controls will want to use the same kind of events. I know having these events as text is not ideal but it does give you the flexibility to trigger anything you can think of. You could have an OnSave, OnFilter, OnDatabind, OnShow, OnDelete, OnOpen, OnError, OnSecondTimeTheUserOpensThis ....  
So a developer should have the ability to expand on the standard events that Qcubed provides.

**Comment by olegabr** [comment]
@matthiaz, what about something like `new QCustomEvent("OnSecondTimeTheUserOpensThis")`?

**Comment by spekary** [comment]
Yes, I am suggesting that you broadcast to all controls. Any controls that care will listen for that particular message. As long as you aren't constantly broadcasting things, this should have very little impact on performance. You should not be broadcasting something that could be a performance bottleneck. It should be a simple messaging service.

All member variables in controls get saved to the form state. So, in your implementation, all the various copies of control references and callbacks will add to the form state. In the implementation I am suggesting, there is no change to the form state, since there are no changes to QControlBase member variables.

JQuery's messaging service is built on top of javascript and basic user interactions. It is complicated, and includes the idea of bubbling of events. I do not think it should be the basis for this feature. You do not want to overload the names of things that are really different items to begin with.

**Comment by vakopian** [comment]
I was not suggesting to use jquery's messaging mechanism nor translating it directly, I was just using it as analogy to think of ideas and ways to do this with a good feature set.

**Comment by matthiaz** [comment]
@olegabr thank you for the idea.
@spekary Ok, I get what you are saying. So I tried converting my code to get read of the class variables.

In QForm I added
```
protected function Broadcast($objSender, $objSignal) {
		foreach($this->objControlArray as $objControl) {
			$objControl->Listen($objSender, $objSignal);
		}
}
```

Then in QControl
```
/**
 * 
 * @param objSignal $objSignal
 * @param array $args
 */
public function Broadcast($objSignal, $args = array()) {
	$_FORM->Broadcast($this, $objSignal, $args);
}
```

In Control A I put
```
$this->Broadcast(new QSaveEvent());
```

And now I should be able to Listen for this Broadcast in Control B. 	
```
	/**
	 * 
	 * @param QControl $objSender
	 * @param QEvent $objSignal
	 */
	public function Listen($objSender, $objSignal) {
		if($objSignal instanceof QSaveEvent) {
			//@TODO: mmm how do I listen for one specific object?
			QApplication::DisplayAlert('Some object saved itself');
		}
	}
```
But here lies the problem. How can I differentiate between objects. I don't want to listen to all save events, I only want to listen to a save event of one specific object. 
And even if I build my if structure in such a way that it only looks at a specific object. It would be hard coded. This means I cannot use the same control in another part of my project without it listening to the event. Maybe I want it to listen to a different event. Or even do something completely different. 

I my first proposal this is easely done by simply assigning another callback to it.

Do tell me if I misunderstood the concept.

**Comment by olegabr** [comment]
@matthiaz, what about something like `new QSaveEvent(array("object" => $theVeryImportantObject))`? And then:
```php
    /**
     * 
     * @param QControl $objSender
     * @param QEvent $objSignal
     */
    public function Listen($objSender, $objSignal) {
        if($objSignal instanceof QSaveEvent && isset($objSignal->data["object"])) {
            QApplication::DisplayAlert('object with id ' . $objSignal->data["object"]->ControlId . ' saved itself');
        }
    }
```

**Comment by matthiaz** [comment]
@Oleg that might work but then the QSaveEvent will have to store the array
somewhere. Which means it will go into the form_state. Unless I
misunderstood, that is what were trying to avoid, right?

2015-02-27 13:31 GMT+01:00 Oleg Abrosimov <notifications@github.com>:

**Comment by spekary** [comment]
So, you want to do dependency injection then. Be able to on-th-fly change what a control listens to, and also how it responds? In most OOP situations, you would subclass the control and override the listener in each situation. The other way is as you said, to save a structure that has the object reference and name of the callback, and put it into the form state. Putting stuff into the form state isn't terrible, you just have to be careful because the data can get stale.

**Comment by olegabr** [comment]
@matthiaz, no, formstate is not used here. you can not emit event in one AJAX callback and respond on it in another one. You always emit and handle it in one AJAX handler. Or you meant something different?

**Comment by spekary** [comment]
Looking through this, I think there are basicly two different implementations being discussed. These seem to be somewhat similar to how BroadcastReceivers and Handlers are implemented in Android programming:

- Broadcasts and broadcast receivers are for general purpose notification of the whole application. You can filter on what broadcasts you are listening to. Android's filter is applied by the system, and you are only sent what you are asking for, but you can always make the listener to the filtering by sending all broadcasts to everyone. For our purposes, it doesn't really matter. Android needs to do it for security reasons.

- Handlers are the ability to give an event queue to a different object. The listener is the one creating the Handler, and it gives it to any object that it wants to broadcast to it. Android does this primarily to manage mult-threading and process issues. Broadcasts are implemented on top of Handlers.

For QCubed's purposes, we can go a lot of directions with this, but we have implementation issues we have to consider. In particular, when do the listeners register, and how dynamic of a process this should be.

If we want to have any kind of object listen to a broadcast, or we want to have a particular listener register with a particular broadcaster, we are going to have to put something in the formstate. We cannot put lambda's and generic functions in the formstate because they are not serializable, but you can put a callable there (an object and a method name essentially).

The simplest way to do this is to have QControls broadcast to all other QControls. Registration is not a problem, because all controls automatically get regsitered. Nothing additional goes in the formstate. Listeners are responsible for filtering. Broadcasters can send a reference to themselves, and a broadcast object which will allow listeners to have enough information to filter.

This method could become slow if someone heavily relied on broadcasting to do work. 

A next step up would be to implement a listener interface, and test which controls implement the interface when they get registered with the form. We would then save the id of that control in a registry at the form level. That would speed things up and have minimal impact on the formstate. Controls would still be required to do their own filtering, as any listener control would receive all broadcasts, but I suspect the impact would be minimal. If its still a problem, then we can add an additional function in the interface that gets called when the control registers with the form, and that would send back the broadcast objects it wants to listen to. That pushes some of the filtering to the form level, making things a little faster, but makes it more cumbersome to be a listener.

Anyway, moving this to v3.2. @matthiaz if you want to try your hand at implementing something like this, please make your pull request against v3.2.
